fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -1292,10 +1295,135 @@ Future<Optional<std::string>> ConsistencyCheckImpl::commit(ReadYourWritesTransac
+		std::sort(keys.begin(), keys.end(), [](const KeyRef& lhs, const KeyRef& rhs) {
+			if (lhs.size() != rhs.size()) {
+				return lhs.size() < rhs.size();
+			}
+			return lhs.compare(rhs) < 0;
+		});
+
+		// Cannot use a range clear because of how keys are ordered in FDB.
+		//   \xff/globalConfig/h/999 -> ...
+		//   \xff/globalConfig/h/1000 -> ...
+		//   \xff/globalConfig/h/1001 -> ...
+		//
+		//   clear_range(\xff/globalConfig/h, \xff/globalConfig/h/1000) results
+		//   in zero key-value pairs being deleted (999 is lexicographically
+		//   larger than 1000, and the range is exclusive).
+		for (int i = 0; i < keys.size() - (kGlobalConfigMaxHistorySize - 1); ++i) {
+			tr.clear(keys[i]);
+		}
+	}
+
+	Arena arena;
+	VectorRef<MutationRef> mutations;
+
+	// Transform writes from special-key-space (\xff\xff/global_config/) to
+	// system key space (\xff/globalConfig/).
+	state RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::Ranges ranges =
+	    ryw->getSpecialKeySpaceWriteMap().containedRanges(specialKeys);
+	state RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::iterator iter = ranges.begin();
+	while (iter != ranges.end()) {
+		Key bareKey = iter->begin().removePrefix(globalConfig->getKeyRange().begin);
+		Key systemKey = bareKey.withPrefix(globalConfigDataPrefix);
+		std::pair<bool, Optional<Value>> entry = iter->value();
+		if (entry.first) {
+			if (entry.second.present()) {
+				mutations.emplace_back_deep(arena, MutationRef(MutationRef::SetValue, bareKey, entry.second.get()));
+				tr.set(systemKey, entry.second.get());
+			} else {
+				mutations.emplace_back_deep(arena, MutationRef(MutationRef::ClearRange, bareKey, keyAfter(bareKey)));
+				tr.clear(systemKey);
+			}
+		}
+		++iter;
+	}
+
+	ProtocolVersion protocolVersion = g_network->protocolVersion();
+
+	// Record the mutations in this commit into the global configuration history.
+	BinaryWriter historyKeyWriter(AssumeVersion(protocolVersion));
+	historyKeyWriter.serializeBytes(globalConfigHistoryPrefix);
+	Key historyKey = addVersionStampAtEnd(historyKeyWriter.toValue());
+
+	BinaryWriter historyMutationsWriter(AssumeVersion(protocolVersion));
+	historyMutationsWriter << mutations;
+
+	tr.atomicOp(historyKey, historyMutationsWriter.toValue(), MutationRef::SetVersionstampedKey);
+
+	// Write version key to trigger update in cluster controller.
+	tr.atomicOp(globalConfigVersionKey,
+	            BinaryWriter::toValue(protocolVersion, AssumeVersion(protocolVersion))
+	                .withPrefix(LiteralStringRef("0123456789")) // placeholder for versionstamp
+	                .withSuffix(LiteralStringRef("\x00\x00\x00\x00")),
+	            MutationRef::SetVersionstampedValue);
+
+	return Optional<std::string>();
+
+}
+
+Future<Optional<std::string>> GlobalConfigImpl::commit(ReadYourWritesTransaction* ryw) {
+	return globalConfigCommitActor(this, ryw);
+}
+
+void GlobalConfigImpl::clear(ReadYourWritesTransaction* ryw, const KeyRangeRef& range) {
+	// TODO
```

Comment:
The current implementation makes this a little bit difficult to implement, because currently each key that should be deleted must be added to the special-key-space write map, so I've held off implementing clear ranges for now.

Reply:
Can you please throw an error? Also, please check with @sfc-gh-clin how this can be done.